### ISSUE TITLE
--nginx fails if original server config is a rewrite

### ISSUE THREAD

**Issue by wizonesolutions** [Neilpang/acme.sh#1115]
Steps to reproduce
------------------
- Add an nginx server configuration on port 80 with the server name for which you are trying to get the certificate. e.g.

```
server {
  listen 80;
  server_name example.com;
  rewrite ^ https://$server_name$request_uri? permanent;
}
```

- Run `--issue` with `--nginx`. It reloads nginx, but Let's Encrypt gets `301` or `403` errors (you can see this in `/var/log/nginx/access.log`.

Workaround: If you instead for example `return "some text";`, that is to say static content, then the rewritten conf file works fine. acme.sh does seem to do the replacement properly in both cases, but perhaps it is not using a strong enough reload command in `--nginx` to overpower `rewrite`? I'm not really sure why it fails, but I know that when I took the `rewrite` out of the original conf, it worked properly...

Just submitting this in case anyone wants to try to reproduce it.

**Comment by LomotHo** [comment]
the same problem

**Comment by LomotHo** [comment]
Now I get a way to solve it, just don't use the rewrite, and let Javascript to do the redirect. 
```nginx
server {
    listen 80;
    server_name www.xxxxxx.com;
    # rewrite ^(.*)$  https://$host$1 permanent;
    root /var/www/www.xxxxxx.com;
    include /etc/nginx/default.d/*.conf;

    location / {
    	index index.html;
    }
}
```
index.html for 80 port
```html
<html>
<script type="text/javascript">
var targetProtocol = "https:";
if (window.location.protocol != targetProtocol)
 window.location.href = targetProtocol +
  window.location.href.substring(window.location.protocol.length);
</script>
</html>
```

**Comment by wizonesolutions** [comment]
I'm using Ansible so I just change back to a rewrite after the cert is
generated successfully. Works ok

16. nov. 2017 18:42 je oseba "lomot" <notifications@github.com> napisala:

**Comment by LomotHo** [comment]
but you will meet the same problem when the certificate is expired, I have just met this. with the rewrite, the acme.sh can't get a new certificate automatically

**Comment by wizonesolutions** [comment]
True. Maybe the hooks could help. Store a non-rewrite version, swap it in
pre and out post.

17. nov. 2017 03:44 je oseba "lomot" <notifications@github.com> napisala:

**Comment by FernandoMiguel** [comment]
You are making this far more complex than it needs to be.

If you already have a cert, you can use tls01 instead of http so validation happens on port 443

If you don't have a cert or have one and prefer validation to happen on port 80, all you need is a location block on your vhost for /.acme

**Comment by FernandoMiguel** [comment]
Here are a few examples 
https://gist.github.com/cecilemuller/a26737699a7e70a7093d4dc115915de8
https://community.letsencrypt.org/t/how-to-nginx-configuration-to-enable-acme-challenge-support-on-all-http-virtual-hosts/5622

**Comment by wizonesolutions** [comment]
What is the proper --issue command to tell acme.sh to use http-01 for
issuance and tls-01 for renewal?

How do I add a proper location block when the verification token it looks
for changes? This is a reverse proxy. There is no docroot.

Sure, one can use standalone (and turn off the whole web server for a bit
whenever acme.sh decides to set up the cron job), but this issue is about a
bug in the nginx plugin.

17. nov. 2017 09:06 je oseba "Fernando Miguel" <notifications@github.com>
napisala:

Here are a few examples
https://gist.github.com/cecilemuller/a26737699a7e70a7093d4dc115915de8
https://community.letsencrypt.org/t/how-to-nginx-configurati
on-to-enable-acme-challenge-support-on-all-http-virtual-hosts/5622

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Neilpang/acme.sh/issues/1115#issuecomment-345173183>,
or mute the thread

**Comment by FernandoMiguel** [comment]
yeah, there's always the option of issuing as standalone, by disabling nginx, and then issuing a new one with TLS01

there is no automation for that.

**Comment by wizonesolutions** [comment]
I went ahead and just switched to using webroot. I moved my `rewrite` to `location /` and added a separate `location /.well-known/acme-challenge/` that I aliased to a specific directory on the machine. I pass that directory in `-w`. I'll leave this open as a bug ticket though.

Ref: https://www.rmedgar.com/blog/using-acme-sh-with-nginx

I didn't do the include thing since I'm using Ansible and can do it with a variable there, but it's basically the same except for also having an https rewrite for non-`/.well-known/acme-challenge/` paths.

**Comment by drlauridsen** [comment]
I know this is a very old thread, but just had the same issue with a 301 https redirect and using --nginx for cert creation.
It probably fails, because the script for the temporary nginx.conf mod ends up adding the acme challenge after the redirect
eg
`server {
    server_name example.com
    return 301 $scheme://www.example.com$request_uri
}`
becomes
`server {
    server_name example.com
    return 301 $scheme://www.example.com$request_uri
    location = /.well-known/acme-challenge/challengeurl{default_type text/plain;return 200 challengeresponse;}
}`

Solution was simply putting the redirect inside a location
`server {
    server_name example.com
    location / {    return 301 $scheme://www.example.com$request_uri }
}`

**Comment by juslintek** [comment]
Hi, same issue in 2023. It add something like that:
```nginx
#ACME_NGINX_START
location ~ "^/\.well-known/acme-challenge/([-_a-zA-Z0-9]+)$" {
  default_type text/plain;
  return 200 "$1.um-kdPjwG6fqLDAMavVp27uEsnRNRDH3TKtAjGHTkNo";
}
#NGINX_START
```

to my server configs:

```nginx
server {
    listen 80;
    listen [::]:80;

    server_name cockpit.domain.com;
#ACME_NGINX_START
location ~ "^/\.well-known/acme-challenge/([-_a-zA-Z0-9]+)$" {
  default_type text/plain;
  return 200 "$1.um-kdPjwG6fqLDAMavVp27uEsnRNRDH3TKtAjGHTkNo";
}
#NGINX_START


    error_log /var/log/nginx/cockpit.error.log warn;
    access_log /var/log/nginx/cockpit.access.log main;

    return 301 https://$host$request_uri;
}
```

And 80 access logs how this:
```log
91.199.212.132 - - [20/May/2023:11:38:58 +0000] "GET /.well-known/acme-challenge/-MzF-Pdckuj5CVN4_UqgiLQvQplLTlNnZe7Z6s8RHkI HTTP/1.1" 301 178 "-" "acme.zerossl.com/v2/DV90" "-"
91.199.212.132 - - [20/May/2023:11:39:01 +0000] "GET /.well-known/acme-challenge/-MzF-Pdckuj5CVN4_UqgiLQvQplLTlNnZe7Z6s8RHkI HTTP/1.1" 301 178 "-" "acme.zerossl.com/v2/DV90" "-"
```

So the location matching fails, because it makes different request, to what is added to nginx configs...

**Comment by neilpang** [comment]
Root cause, as @drlauridsen and @juslintek pinpointed: a `rewrite` / `return 301` at the **server** level runs in nginx's rewrite phase, before location matching. `--nginx` only inserts a `location` block for the challenge -- it can't override a server-level redirect, so Let's Encrypt's request to `/.well-known/acme-challenge/...` is 301'd away before the challenge location is reached. Auto-neutralizing a server-level `return`/`rewrite` would be fragile, so the `--nginx` patch doesn't attempt it.

Workaround -- move the redirect from the server block into `location /`:
```nginx
server {
    listen 80;
    server_name example.com;
    location / { return 301 https://$host$request_uri; }
}
```
The challenge location acme.sh inserts (`location ~ ^/\.well-known/acme-challenge/...`) is a regex match, which outranks the `location /` prefix, so it wins for the challenge path while everything else still redirects. Or skip `--nginx` and use `--webroot` (with a dedicated acme-challenge location) or a DNS mode.
